Issue 228 - remove and warn unused selex patterns

k-doering-NOAA opened the pull request:
## What issue(s) does this PR address? Describe and add issue numbers, if applicable.

#228 

This pull requests adds or edits warnings and exit statements for selectivity patterns that have been discontinued or are not recommended.

- [x] Link issue on pull request

## What tests have been done? Upload any model input files created for testing in a zip file, if possible.

I compiled and tested out all selectivity patterns that have exits and warnings or just warnings (length patterns 2, 4, 7, 30, 8; age patterns 13, 18).  I did this by modifying the selectivity patterns and number of lines required as in the manual (note pattern says it need 8 parameter lines, but only needs 6)

## What tests/review still need to be done? Who can do it, and by when is it needed (ideally)?

It would be good for Rick to review the code. In particular, if the single warnings when the model run is not supposed to exit on error are implemented correctly and if there are any changes in wording needed.

Someone checking at least a few of the patterns would be helpful, to make sure my tests also work elsewhere. 

## Check which is true. This PR requires:

- [ ] Changes in r4ss
- [x] Changes to SS3 manual
- [x] Changes to SSI (the SS3 GUI)
- [x] An entry in the stock synthesis change log (new features, bug reports)

##  Describe any changes in r4ss/SS3 manual/SSI/change log that are needed (if checked):
### for ss3 manual

It might be good to remove the discontinued patterns from the manual; I think they have all been discontinued for years at this point. 

### for change log
Add warnings for discontinued selectivity patterns

iantaylor-NOAA commented:
I like the idea of removing the discontinued patterns from the User Manual. If someone has a very old model that includes one of these patterns, they can use a very old manual for information about it.

Rick-Methot-NOAA commented:
Did you check & update the list of selex fxns that displays in control.ss_new?

k-doering-NOAA commented:
> Did you check & update the list of selex fxns that displays in control.ss_new?

No, I didn't, good point! I can make that change.

Rick-Methot-NOAA reviewed (changes requested):
Let me try to hunt the history of:
  //  SS_Label_Info_22.3.2 #case 2 discontinued; use pattern 8 for double logistic
  //  SS_Label_Info_22.3.24 #case 24 size selectivity using double_normal_plateau and lots of bells and whistles
I suspect that I needed an experimental version of 24, so I copied the 24 code, renamed it as case 2.  So, I will do a code compare of the 24(2)  relative to the actual 24.

k-doering-NOAA commented:
Ok @Rick-Methot-NOAA, let me know what you find out!

Rick-Methot-NOAA commented:
all that needs to change are the comments labelling the code:
this line gets deleted:
// SS_Label_Info_22.3.2 #case 2 discontinued; use pattern 8 for double logistic
this line:
// SS_Label_Info_22.3.24 #case 24 size selectivity using double_normal_plateau and lots of bells
becomes:
// SS_Label_Info_22.3.2 #case 2 is a modification of case24 size selectivity for better performance with sex-specific

k-doering-NOAA commented:
OK. Actually, were you suggesting we keep two selectivity patterns,  or just the one with better performance (e.g., should actually 24 be deprecated and 2 kept)? Or maybe we just continue to say 2 is deprecated and add the small chunk of code you discovered to 24 to improve the pattern?

Rick-Methot-NOAA commented:
Keep two patterns for continued backward compatibility:
  case 24
 24 modified and currently with case 2 as its reference.

iantaylor-NOAA commented:
Linking in issue #134 for future reference, which explains the origins of the current selectivity pattern 2.

I like the idea of making 24 be the one with the modification, if I'm interpreting that suggestion correctly, and 2 be the backup for compatibility if we can document it adequately so that people know to look there.

This only matters for the small subset of models which had sex-specific double-normal selectivity which was estimated as dome-shaped.

Rick-Methot-NOAA commented:
I like Ian's suggestion.

k-doering-NOAA commented:
This does make sense, maybe we just have a note appear in the warning file if using 24?

It makes sense to have the improved version be added to 24. I will work on making these changes.

Rick-Methot-NOAA commented:
warning would be good:
IF using sex-specific version of 24, then warn user to consider better performance of case 2.

iantaylor-NOAA commented:
Warning could be conditional on this check for selex type 2 or 24 within the check for sex-specific selectivity: https://github.com/nmfs-stock-synthesis/stock-synthesis/blob/cdac3a738f01e2b9040546f908e38ea06e7994ea/SS_readcontrol_330.tpl#L3027-L3034

It seems like we're talking about two potential options:
1. **restore selex type 2** (removed in this pull request because we forgot to document that it has a purpose) **and leave selex type 24 as it stands with a warning** that selectivity type 2 could be used if you want the estimate of scale to apply to the final selectivity as well as the plateau, rather than just the plateau. This has the benefit of providing easy backwards compatibility and avoids explaining why models may have changed.
2. **restore selex type 2 and then swap the code for 2 and 24** so that type 2 matches the current type 24 where the scale parameter only applies to the plateau and not the final, and type 2 would be used for backwards compatibility. This has the benefit of bringing all the existing models to the "improved" parameterization automatically (if we agree that it's an improvement) without having to nudge people to explore that change. However, it has the drawback that some people may be surprised to see a model change and have to go through extra work to understand why.

I don't have any strong opinion on which is best.

Rick-Methot-NOAA commented:
Thanks for the clear description Ian.  Approach 1 would mean that a user wanting to keep the existing selex 24 for a sex-specific selectivity would get the warning every time until they switch to the new 2.
With approach 2, a user (there are very, very few!) of current sex-specific 24 would see a change in result and would not get a warning.  Then they write to us and we explain.  That seems OK.

k-doering-NOAA commented:
Ok, I went with option 2. I also realized the 2 patterns only had one line of difference, so combined the code for both in the switch statement and added an if/else statement.

I used tested the change using a modified version of big skate (had to change parameter 6 from -999 to 0.01 test the code). I also modified it 1 additional time to add pattern 2 instead of 24 and checked with print debugging that SS3 was going to the correct if statements.

This same test could be done, looking for a warning message when using pattern 2 and none when using pattern 24; I imagine there would be a difference in the selectivity patterns when using these.

Rick-Methot-NOAA commented:
I like your approach to coding this.  I think it is ready to merge, then get documented in the change log and the manual and SSI.

Rick-Methot-NOAA reviewed (commented):
all good.  merge OK
